### PR TITLE
Remove "unreachable" instruction from LLVM IR for SYCL devices

### PR DESCRIPTION
**Pull request by srividya-sundaram**
C/C++ standard functions with `__attribute__ ((__noreturn__))` in SYCL device code emit "unreachable" instruction in LLVM IR. This PR removes the "ur" instruction in LLVM IR for SYCL devices.

### CODE REVIEW

File: clang/lib/CodeGen/CGCall.cpp [hunk 1]
```diff
@@ -5017,8 +5017,9 @@ RValue CodeGenFunction::EmitCall(const CGFunctionInfo &CallInfo,
         EmitNounwindRuntimeCall(Fn);
       }
     }
-
-    EmitUnreachable(Loc);
+    // Do not emit unreachable instruction for Intel NEO driver and Gen "X" GPUs
+    if (getTarget().getTriple().getSubArch() != llvm::Triple::SPIRSubArch_gen)
```
Comment (Fznamznon) [inline clang/lib/CodeGen/CGCall.cpp:5021]: I don't get why NEO driver can't handle this.
It seems that SPIR-V has unreachable instruction as well.

Comment (asavonic) [inline clang/lib/CodeGen/CGCall.cpp:5021]: I'm not sure that we set `SPIRSubArch_gen` for generic SPIR-V compilation (JIT). Can we disable `noreturn` for all SPIR targets?

Comment (erichkeane) [reply clang/lib/CodeGen/CGCall.cpp:5021]: Should this condition be specific to SYCL devices?  I'm not sure its appropriate to set it for all other offloads, I'm not sure OpenMP or CUDA/etc would necessarily want this change.

**Comment by erichkeane** [comment]
@savonic: Mind taking 1 more look?  I'm going to approve, but I want your final thought as well.

File: clang/lib/CodeGen/CGCall.cpp (hunk 1 above)
Comment (bader) [reply clang/lib/CodeGen/CGCall.cpp]: > SPIR-V does not support semantic of `noreturn` functions; there is no (portable) way to terminate an invocation or jump out of a function.

How is this related to removing "unreachable" instruction from LLVM IR?

As noted by @Fznamznon, SPIR-V has ["unreachable"](https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#OpUnreachable) operation, which behavior is undefined by the spec. Could you elaborate more on why you think that removing "unreachable" is the right way to fix the original problem? I don't see the connection.

**Comment by erichkeane** [comment]
> 
> 
> > C/C++ standard functions with `__attribute__ ((__noreturn__))` in SYCL device code emit "unreachable" instruction in LLVM IR. This "unreachable" instruction causes Intel NEO driver and Gen "X" GPUs to hang.
> 
> Isn't this expected behavior? Maybe I don't understand the behavior of this attribute it looks like if I call "noreturn" function, it might run forever. Right?

Well, simply that the function doesn't 'return', and the optimizer is allowed to assume that it is a 'terminate' type of thing.  Typically, a noreturn function must terminate the program (as never-ending loops are considered UB).

I forget the context of the bug report, but my understanding is that the driver folks didn't have a mechanism for handling this in the case of abort.  Presumably, they would need to teach said drivers to abort, but I thought there wasn't a good way to do so.

File: clang/lib/CodeGen/CGCall.cpp [hunk 2]
```diff
@@ -5017,8 +5017,9 @@ RValue CodeGenFunction::EmitCall(const CGFunctionInfo &CallInfo,
         EmitNounwindRuntimeCall(Fn);
       }
     }
-
-    EmitUnreachable(Loc);
+    // Emit unreachable only for non-sycl devices
+    if (getTarget().getTriple().getSubArch() != getLangOpts().SYCLIsDevice)
```
Comment (Fznamznon) [inline clang/lib/CodeGen/CGCall.cpp:5021]: Wow, I guess it should look like this
```suggestion
    if (!getLangOpts().SYCLIsDevice)
```

Comment (Srividya2212) [reply clang/lib/CodeGen/CGCall.cpp:5021]: Sorry, I just focused on changing the spir_gen part. Will fix it.

**Comment by bader** [comment]
> > > C/C++ standard functions with `__attribute__ ((__noreturn__))` in SYCL device code emit "unreachable" instruction in LLVM IR. This "unreachable" instruction causes Intel NEO driver and Gen "X" GPUs to hang.
> > 
> > 
> > Isn't this expected behavior? Maybe I don't understand the behavior of this attribute it looks like if I call "noreturn" function, it might run forever. Right?
> 
> Well, simply that the function doesn't 'return', and the optimizer is allowed to assume that it is a 'terminate' type of thing. Typically, a noreturn function must terminate the program (as never-ending loops are considered UB).
> 
> I forget the context of the bug report, but my understanding is that the driver folks didn't have a mechanism for handling this in the case of abort. Presumably, they would need to teach said drivers to abort, but I thought there wasn't a good way to do so.

Is code comments below captures it right?
```c++
void foo() { while(1); } // UB
[[noreturn]] void foo() { while(1); } // must terminate - not UB
```

**Comment by erichkeane** [comment]
> 
> 
> > > > C/C++ standard functions with `__attribute__ ((__noreturn__))` in SYCL device code emit "unreachable" instruction in LLVM IR. This "unreachable" instruction causes Intel NEO driver and Gen "X" GPUs to hang.
> > > 
> > > 
> > > Isn't this expected behavior? Maybe I don't understand the behavior of this attribute it looks like if I call "noreturn" function, it might run forever. Right?
> > 
> > 
> > Well, simply that the function doesn't 'return', and the optimizer is allowed to assume that it is a 'terminate' type of thing. Typically, a noreturn function must terminate the program (as never-ending loops are considered UB).
> > I forget the context of the bug report, but my understanding is that the driver folks didn't have a mechanism for handling this in the case of abort. Presumably, they would need to teach said drivers to abort, but I thought there wasn't a good way to do so.
> 
> Is code comments below captures it right?
> 
> ```c++
> void foo() { while(1); } // UB
> [[noreturn]] void foo() { while(1); } // must terminate - not UB
> ```

Infinite loops themselves are UB, see http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1528.htm

By my reading of the standard, BOTH while loops there are UB.  I don't believe that the noreturn attribute changes that at all.  It simply means that the compiler can optimize assuming that code after the call to a no-return function never executes.

File: clang/lib/CodeGen/CGCall.cpp (hunk 1 above)
Comment (asavonic) [reply clang/lib/CodeGen/CGCall.cpp]: If a function is marked as `noreturn`, it has to either terminate, throw or longjmp. Otherwise the behavior is undefined, so clang emits an unreachable instruction after a call to a noreturn function.

There is no way to "not return" in SPIR-V, so any call to a `noreturn` function is a UB. Currently this UB effectively marks the whole basic block unreachable. The proposed patch changes this UB to ignore `noreturn`. This gives more room to handle existing code that has `noreturn` functions.

Comment (bader) [reply clang/lib/CodeGen/CGCall.cpp]: Why can't we leave SPIR-V unreachable and let GPU back-end to handle UB? If I understand correctly SPIR-V already ignores `noreturn` attribute.

Comment (asavonic) [reply clang/lib/CodeGen/CGCall.cpp]: Frontend has more information and can distinguish `noreturn` attribute from other uses of unreachable (the corresponding builtin, for instance). 
Also, when optimizations are enabled in the frontend, optimizer may remove code related to an unreachable.

Comment (Srividya2212) [reply clang/lib/CodeGen/CGCall.cpp]: @asavonic I think this change is causing the original test you added (devicelib/assert.cpp) to fail.
[assert.cpp](http://ci.llvm.intel.com:8010/#/builders/2/builds/2826)
`Basic Block in function '_ZZZ11simple_vaddIiLm3EEvRKSt5arrayIT_XT0_EES4_RS2_ENKUlRN2cl4sycl7handlerEE_clES9_ENKUlNS7_2idILi1EEEE_clESC_' does not have terminator!`

Can you and @bader decide and suggest what's the best way to address the original issue of this PR?

Comment (asavonic) [reply clang/lib/CodeGen/CGCall.cpp]: > @asavonic I think this change is causing the original test you added (devicelib/assert.cpp) to fail.
> [assert.cpp](http://ci.llvm.intel.com:8010/#/builders/2/builds/2826)
> `Basic Block in function '_ZZZ11simple_vaddIiLm3EEvRKSt5arrayIT_XT0_EES4_RS2_ENKUlRN2cl4sycl7handlerEE_clES9_ENKUlNS7_2idILi1EEEE_clESC_' does not have terminator!`

Sounds like we need to skip the entire block at `if (CI->doesNotReturn())`.

**Comment by asavonic** [comment]
Please align the PR title with the latest revision of the patch. It is no longer limited to Intel NEO driver.

File: sycl/test/devicelib/assert.cpp [hunk 3]
```diff
@@ -137,11 +134,10 @@ void simple_vadd(const std::array<T, N> &VA, const std::array<T, N> &VB,
   }
   if (unsupported && getenv("SKIP_IF_NO_EXT")) {
     fprintf(stderr, "Device has no support for cl_intel_devicelib_assert, "
-                    "skipping the test\n");
+                    "skipping the test \n");
```
Comment (alexbatashev) [inline sycl/test/devicelib/assert.cpp:137]: Why is this change required?
